Avatar Git Settings Popover

tidy-dev opened the pull request:
xref: https://github.com/github/accessibility-audits/issues/3012

## Description
This makes clicking the avatar show the users git configuration info and provide a means to update if need be.

### Screenshots

https://user-images.githubusercontent.com/75402236/236016641-d12e6a26-02e5-4924-b66e-65573c45ce31.mp4



## Release notes
Notes: [Improved] Adds committing avatar popover to see git configuration and ability to open git configuration settings.

tidy-dev committed:
Add committing as popover

tidy-dev committed:
Add checking local/global config and buttons instead of links

tidy-dev committed:
Aria label depending on if commit is misattributed

tidy-dev committed:
Update button text to be more accurate

tidy-dev committed:
Fix property readonlyness

niik commented on app/src/ui/changes/commit-message.tsx:
```diff
@@ -423,6 +424,13 @@ export class CommitMessage extends React.Component<
     }
   }
 
+  private isGitConfigLocal = async () => {
+    const { repository } = this.props
+    const localName = await getConfigValue(repository, 'user.name', true)
+    const localEmail = await getConfigValue(repository, 'user.email', true)
+    return localName !== null || localEmail !== null
+  }
+
```
Could we move this inside of the CommitMessageAvatar component instead? I don't believe we've ever used the pattern of passing a promise to a component before and as this is written now I believe (haven't tested but sure looks like it) we'll end up invoking this method in every render which could significantly affect performance.

I'd rather we ran this when the `CommitMessageAvatar` component is mounted and then optionally whenever the app regains focus similar to to https://github.com/desktop/desktop/blob/44429be2eba381d1f4abd882cfc818baf711a46a/app/src/ui/add-repository/create-repository.tsx#L674-L678

tidy-dev replied:
In the commit avatar component, I already had logic to run this in the constructor and then on `componentDidUpdate` if the user name or email had changed as I figured that was a decent indication of when git configuration changed (didn't know that having in the props would cause it to run).. tho this timing is not fool proof as someone could have the same name and email in global and in local. I don't think on app focus would be frequent enough because I need the change to pick up if the user goes to the repository settings and modifies them so that the avatar shows the correct message.

tidy-dev replied:
I don't think it is that important for this to be fool proof as it is just whether the user is directed to the repository settings or git user settings.. and to be honest, I doubt this will be a high traffic area to the extent it is used (people don't change their git config location often) it will likely be accurate. But, if anyone feels it should be better.. we could track when the git config location changes on the repository state to send down to the avatar component.

tidy-dev committed:
Move git location check into avatar component

tidy-dev committed:
Reduce times this runs

Evidently the user object gets replaced frequently so this is running many times based on the object reference changing. This reduces it to just on user name or email change

sergiou87 reviewed (approved):
Code looks good and works as expected 💖 

Only ran into a bug that was already before these changes: when the "misattributed commit" warning is generated by the email set locally in the repo, the popover mentions global config and all that 🤔 I can't remember if it was always like that tbh 🤯 

I can take care of that in a different PR!

sergiou87 reviewed (dismissed):
Code looks good and works as expected 💖 

Only ran into a bug that was already before these changes: when the "misattributed commit" warning is generated by the email set locally in the repo, the popover mentions global config and all that 🤔 I can't remember if it was always like that tbh 🤯 

I can take care of that in a different PR!

tidy-dev committed:
Merge branch 'development' into avatar-git-settings-popover

tidy-dev committed:
Remove double header

tidy-dev committed:
Position tweaking between misattributed and regular popover

tidy-dev commented:
After merge update from misattributed commit fixed positioning, I had to do some tweaking so that misattributed versus git author popover were positioned correctly. But, I think that is all that needed updating.

sergiou87 reviewed (changes requested):
After the last changes, positioning doesn't seem correct… unless I do something that triggers a layout?

https://github.com/desktop/desktop/assets/1083228/22a5d0c8-da0d-4267-80f2-d3e14d337646

tidy-dev committed:
Fix default height

tidy-dev committed:
Different default for different popover sizes

tidy-dev committed:
Don't default to showing "Other" while loading in git configs to lessen chance of jumpy screen

tidy-dev committed:
Pass loader in on repo settings too

tidy-dev committed:
Remove jumpiness of Git and Repo Git settings when misattributed warning is present

sergiou87 reviewed (approved):
As discussed, my PR https://github.com/desktop/desktop/pull/16717 will take care of the positioning issues, other than that looks great!

Also thank you for taking care of the git preferences state issue + jumpiness 🙇‍♂️  Looks so much better now 🤩 

:shipit: ❗
